src/proc/typifier.rs

```diff
@@ -457,6 +474,15 @@ impl Typifier {
                     size,
                     width,
                 } => Resolution::Value(crate::TypeInner::Vector { kind, size, width }),
+                crate::TypeInner::Matrix {
+                    columns,
+                    rows,
+                    width,
+                } => Resolution::Value(crate::TypeInner::Matrix {
```

Comment:
does this actually work anywhere? I thought type casts are limited to vectors and scalars

Reply:
You're right, this results in `OpBitcast` instructions for `OpTypeMatrix` which is invalid. I think I'll just take that change out of this PR and look for another way to make the shader in #357 parse properly later. Would that be alright?

Reply:
That would be fine, yes.
Could you point me to where exactly #357 needs this cast?

Reply:
If I interpret everything right it's in the line `mat3 normal_matrix = mat3(transpose(inverse(model_matrix)));` and seems to be at the `mat3(` or the assignment step.

The log output is as follows:

```
[2021-02-14T17:12:38Z DEBUG naga::proc::typifier] Resolving Handle(16) = AccessIndex { base: Handle(14), index: 1 } : Handle(Handle(1))
[2021-02-14T17:12:38Z DEBUG naga::proc::typifier] Resolving Handle(17) = Math { fun: Inverse, arg: Handle(13), arg1: None, arg2: None } : Value(Matrix { columns: Quad, rows: Quad, width: 4 })
[2021-02-14T17:12:38Z DEBUG naga::proc::typifier] Resolving Handle(18) = Math { fun: Transpose, arg: Handle(17), arg1: None, arg2: None } : Value(Matrix { columns: Quad, rows: Quad, width: 4 })
thread 'main' panicked at 'ParseError { kind: SemanticError("Can\'t resolve type: IncompatibleOperand { op: \"as\", operand: \"Matrix { columns: Quad, rows: Quad, width: 4 }\" }") }', examples\convert.rs:54:31
```

Reply:
ah, so you are talking about extracting mat3 out of mat4. We don't have IR for this yet. We need to figure out how upstream WebGPU is going to do this.
You can always work around this though (in GLSL frontend, for example), by doing `mat3(matrix.x.xyz, matrix.y.xyz, matrix.z.xyz)`